Allow to specify Go version

The linter github action doesn't allow specifying a Go version. It overwrtes it to `1` in https://github.com/golangci/golangci-lint-action/blob/1d3f25b/src/install.ts#L60.

It would be very useful if the linter picked the Go version specified in go.mod or at least had an action parameter. I made a working PoC in https://github.com/x1ddos/golangci-lint-action/commit/6f8897f but it's for the action version 1.2.2.

We have a project which can't upgrade to the latest go1.15 and had to remove the action in https://github.com/digitalbitbox/bitbox-wallet-app/pull/982.

Yeah, good point about minimal supported version. So, maybe just allowing to specify `go-version` parameter would work best in the end.

 > Based on the intent, a module with go 1.14 in the go.mod file should work on 1.15.

It should but sometimes doesn't in practice. Here's some output from running the linter on https://github.com/digitalbitbox/bitbox-wallet-app.

```
$ go version
go version go1.14.7 linux/amd64

$ golangci-lint --version
golangci-lint has version v1.27.0 built from (unknown, mod sum: "h1:VYLx63qb+XJsHdZ27PMS2w5JZacN0XG8ffUwe7yQomo=") on (unknown)

$ go clean -cache
$ golangci-lint run
$ echo $?
0
```

```
$ go version
go version go1.15 linux/amd64

$ golangci-lint --version
golangci-lint has version v1.27.0 built from (unknown, mod sum: "h1:VYLx63qb+XJsHdZ27PMS2w5JZacN0XG8ffUwe7yQomo=") on (unknown)

$ go clean -cache
$ golangci-lint run
level=warning msg="[runner] Can't run linter goanalysis_metalinter: goprintffuncname: failed prerequisites: [inspect@github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server: analysis skipped: errors in package: [src/github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server/server.go:40:8: could not import C (cgo preprocessing failed)]]"
level=warning msg="[runner] Can't run linter unused: buildir: analysis skipped: errors in package: [src/github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server/server.go:40:8: could not import C (cgo preprocessing failed)]"
level=error msg="Running error: buildir: analysis skipped: errors in package: [src/github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server/server.go:40:8: could not import C (cgo preprocessing failed)]"
```

```
$ go version
go version go1.14.7 linux/amd64

$ golangci-lint --version
golangci-lint has version v1.30.0 built from (unknown, mod sum: "h1:UhdK5WbO0GBd7W+k2lOD7BEJH4Wsa7zKfw8m3/aEJGQ=") on (unknown)

$ go clean -cache
$ golangci-lint run
WARN [linters context] Panic: exhaustive: package "btc" (isInitialPkg: false, needAnalyzeSou[7/1076]
e): interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident: goroutine 17548 [running]:
runtime/debug.Stack(0x14f69cf, 0x3c, 0xc00266b290)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1(0xc001b63560)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:508 +0x1b5
panic(0x13157e0, 0xc012a2c810)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/nishanths/exhaustive.missingCasesTextEdit(0xc00064c0c0, 0xc007c51000, 0xc009a8a600, 0xc01
6872810, 0xc0047f69b0, 0xc012a2c750, 0x0, 0x0, 0x0, 0x0, ...)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:312 +0x873
github.com/nishanths/exhaustive.computeFix(0xc00f278000, 0xc00064c0c0, 0xc007c51000, 0xc016872810, 0
xc0047f69b0, 0x0, 0xc012a2c750, 0x0, 0x0, 0x0, ...)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:174 +0xd8
github.com/nishanths/exhaustive.reportSwitch(0xc00f278000, 0xc016872810, 0xc00c304400, 0xc0047f69b0,
 0xc012a2c750, 0x0, 0xc007c51000)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:146 +0x53d
github.com/nishanths/exhaustive.checkSwitchStatements.func1(0x1683e20, 0xc016872810, 0x1, 0xc004032c
00, 0x4, 0x20, 0x1)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:117 +0x474
golang.org/x/tools/go/ast/inspector.(*Inspector).WithStack(0xc00b2b5d40, 0xc002746c20, 0x1, 0x1, 0xc
00266bc30)
        pkg/mod/golang.org/x/tools@v0.0.0-20200724022722-7017fd6b1305/go/ast/inspector
/inspector.go:126 +0x13f
github.com/nishanths/exhaustive.checkSwitchStatements(0xc00f278000, 0xc00b2b5d40, 0xc002746cb8)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:24 +0xd6
github.com/nishanths/exhaustive.run(0xc00f278000, 0x51354fe14, 0x1f1a600, 0xc00fb9aa70, 0x2)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/exh
austive.go:170 +0x111
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc001b63560)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:590 +0xa25
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:512 +0x2a
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc0012905a0, 0x1477177, 0xa
, 0xc0031f6770)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/timeutils/stopwatch.go:1
11 +0x50
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0xc001b63560)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:511 +0x91
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc00b855
ef0, 0xc001b63560)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:1059 +0x61
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:1054 +0x306
```

```
$ go version
go version go1.15 linux/amd64

$ golangci-lint --version
golangci-lint has version v1.30.0 built from (unknown, mod sum: "h1:UhdK5WbO0GBd7W+k2lOD7BEJH4Wsa7zKfw8m3/aEJGQ=") on (unknown)

$ go clean -cache
$ golangci-lint run
WARN [linters context] Panic: exhaustive: package "btc" (isInitialPkg: false, needAnalyzeSource: tru
e): interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident: goroutine 18546 [running]:
runtime/debug.Stack(0x14f69cf, 0x3c, 0xc001d67290)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1(0xc002eef680)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:508 +0x1b5
panic(0x13157e0, 0xc003f023f0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/nishanths/exhaustive.missingCasesTextEdit(0xc0013840c0, 0xc006b45380, 0xc004410400, 0xc00
1cdbda0, 0xc00185eaa0, 0xc003f02330, 0x0, 0x0, 0x0, 0x0, ...)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:312 +0x873
github.com/nishanths/exhaustive.computeFix(0xc006418780, 0xc0013840c0, 0xc006b45380, 0xc001cdbda0, 0
xc00185eaa0, 0x0, 0xc003f02330, 0x0, 0x0, 0x0, ...)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:174 +0xd8
github.com/nishanths/exhaustive.reportSwitch(0xc006418780, 0xc001cdbda0, 0xc004402300, 0xc00185eaa0,
 0xc003f02330, 0x0, 0xc006b45380)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:146 +0x53d
github.com/nishanths/exhaustive.checkSwitchStatements.func1(0x1683e20, 0xc001cdbda0, 0x1, 0xc0027ad2
00, 0x4, 0x20, 0x1)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:117 +0x474
golang.org/x/tools/go/ast/inspector.(*Inspector).WithStack(0xc0060cc160, 0xc003852c20, 0x1, 0x1, 0xc
001d67c30)
        pkg/mod/golang.org/x/tools@v0.0.0-20200724022722-7017fd6b1305/go/ast/inspector
/inspector.go:126 +0x13f
github.com/nishanths/exhaustive.checkSwitchStatements(0xc006418780, 0xc0060cc160, 0xc003852cb8)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/swi
tch.go:24 +0xd6
github.com/nishanths/exhaustive.run(0xc006418780, 0x3786740b7, 0x1f1a600, 0xc009bb7360, 0x2)
        pkg/mod/github.com/nishanths/exhaustive@v0.0.0-20200708172631-8866003e3856/exh
austive.go:170 +0x111
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc002eef680)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:590 +0xa25
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:512 +0x2a
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc001729c70, 0x1477177, 0xa
, 0xc0019c8f70)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/timeutils/stopwatch.go:1
11 +0x50
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0xc002eef680)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/run
ner.go:511 +0x91
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc0061ca
b50, 0xc002eef680)
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/runner.go:1059 +0x61
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze
        pkg/mod/github.com/golangci/golangci-lint@v1.30.0/pkg/golinters/goanalysis/runner.go:1054 +0x306
WARN [runner] Can't run linter unused: buildir: analysis skipped: errors in package: [src/github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server/server.go:40:8: could not import C (cgo preprocessing failed)]
ERRO Running error: buildir: analysis skipped: errors in package: [src/github.com/digitalbitbox/bitbox-wallet-app/frontends/qt/server/server.go:40:8: could not import C (cgo preprocessing failed)]
```

Even if it did work, the project we're working on can't officially support go1.15 due to some dependencies like gomobile and gopherjs.

--- a/README.md
+++ b/README.md
@@ -47,6 +47,9 @@ jobs:
 
           # Optional: show only new issues if it's a pull request. The default value is `false`.
           # only-new-issues: true
+
+          # Optional: if set to true then the action will use pre-installed Go
+          # skip-go-installation: true
 ```
 
 We recommend running this action in a job separate from other jobs (`go test`, etc)
@@ -94,7 +97,7 @@ jobs:
 You will also likely need to add the following `.gitattributes` file to ensure that line endings for windows builds are properly formatted:
 
 ```.gitattributes
-*.go text eol=lf 
+*.go text eol=lf
 ```
 
 ## Comments and Annotations

--- a/action.yml
+++ b/action.yml
@@ -1,4 +1,3 @@
----
 name: "Run golangci-lint"
 description: "Official golangci-lint action with line-attached annotations for found issues, caching and parallel execution."
 author: "golangci"
@@ -21,7 +20,10 @@ inputs:
     description: "if set to true and the action runs on a pull request - the action outputs only newly found issues"
     default: false
     required: true
-
+  skip-go-installation:
+    description: "if set to true then action uses pre-installed Go"
+    default: false
+    required: true
 runs:
   using: "node12"
   main: "dist/run/index.js"

--- a/src/install.ts
+++ b/src/install.ts
@@ -56,6 +56,12 @@ export async function installLint(versionConfig: VersionConfig): Promise<string>
 }
 
 export async function installGo(): Promise<void> {
+  const skipGoInstallation = core.getInput(`skip-go-installation`, { required: true }).trim()
+  if (skipGoInstallation.toLowerCase() == "true") {
+    core.info(`Skipping the installation of Go`)
+    return
+  }
+
   const startedAt = Date.now()
   process.env[`INPUT_GO-VERSION`] = `1`
   await setupGo()

--- a/src/run.ts
+++ b/src/run.ts
@@ -187,6 +187,7 @@ async function runLint(lintPath: string, patchPath: string): Promise<void> {
 export async function run(): Promise<void> {
   try {
     const { lintPath, patchPath } = await core.group(`prepare environment`, prepareEnv)
+    core.addPath(path.dirname(lintPath))
     await core.group(`run golangci-lint`, () => runLint(lintPath, patchPath))
   } catch (error) {
     core.error(`Failed to run: ${error}, ${error.stack}`)

